docs/fsharp/tour.md

```diff
@@ -1,4 +1,158 @@
+
+## Recursive Functions
+
+Processing collections or sequences of elements is typically done with [recursion](language-reference/functions/index.md#recursive-functions) in F#.  Although F# has support for loops and imperative programming, recursion is preferred because it is easier to guarantee correctness.
+
+>![NOTE]
+This same makes use of the pattern matching via the `match` expression.  This fundamental construct is covered later in this article.
+
+[!code-fsharp[RecursiveFunctions](../../samples/snippets/fsharp/tour.fs#L323-L349)]
+
+F# also has full support for "tail call optimization", which for the uninitiated is a way to optimize recursive calls so that they are just as fast as a loop construct.  This is fairly unique, as most languages require implementing things like Stack Trampolines to avoid overflowing the program Stack with recursive calls.
+
+## Record and Discriminated Union Types
+
+Record and Union types are two fundamental data types used in F# code, and are generally the best way to represent data in an F# program.  Although this makes them similar to classes in other languages, one of their primary differences is that they have structural equality semantics.  This means that they are "natively" comparable and equality is straightforward - just check if one is equal to the other.
+
+[Records](language-reference/records.md) are an aggregate of named values, with optional members (such as methods).  If you're familiar with C# or Java, then these should feel similar to POCOs or POJOs - just with less ceremony and structural equality.
+
+[!code-fsharp[Records](../../samples/snippets/fsharp/tour.fs#L357-L373)]
+
+[Discriminated Unions](language-reference/discriminated-unions.md) are values which could be a number of named forms or cases.  Data stored in the type can be one of several distinct values.
+
+[[!code-fsharp[Unions](../../samples/snippets/fsharp/tour.fs#L381-L432)]
+
+You can also use Unions as *single-case unions*, to help with domain modeling over primitive types.  Often times, strings and other primitive types are used to represent types of information that shouldn't be interchangeable.  However, using only the string representation can lead to that mistake quite easily!  Representing each type of information as a distinct single-case union can enforce correctness in this scenario.
+
+[[!code-fsharp[Unions](../../samples/snippets/fsharp/tour.fs#L434-L437)]
+
+Additionally, Discriminated Unions also support recursive definitions, allowing you to easily represent trees and inherently recursive data.  For example, here's how you can represent a Binary Search Tree with `exists` and `insert` functions.
+
+[[!code-fsharp[Unions](../../samples/snippets/fsharp/tour.fs#L439-L466)]
+
+Because Discriminated Unions allow you to represent the recursive structure of the tree in the data type, operating on it is very straightforward and nets you correctness guarantees with pattern matching.  See below for more.
```

Comment:
"nets you correctness guarantees" sounds awkward. How about reversing the clauses:

"Operating on this recursive structure is straightforward and guarantees correctness. It also supports pattern matching, as shown below.